cmd/node/main.go

```diff
@@ -85,12 +73,132 @@ func main() {
+
+	// egCtx is cancelled if any function called with eg.Go() returns an error.
+	eg, egCtx := errgroup.WithContext(ctx)
 
-	server, err := api.NewAPIServer(router, log, proxy.Rules, nodeConfig)
+	proxy := reverse_proxy.NewProxyServer(log, nodeConfig)
+	tlsConfig, err := nodeConfig.TLSConfig()
 	if err != nil {
 		log.Fatal().Err(err)
 	}
-	defer server.Close()
-	err = server.ListenAndServeTLS(nodeConfig.NodeCertPath(), nodeConfig.NodeKeyPath())
-	log.Fatal().Msgf("Habitat API server error: %s", err)
+	proxyServer := &http.Server{
+		Addr:        fmt.Sprintf(":%s", constants.DefaultPortReverseProxy),
+		TLSConfig:   tlsConfig,
+		Handler:     proxy,
+		BaseContext: func(net.Listener) context.Context { return context.Background() },
+	}
+
+	eg.Go(serveFn(proxyServer, "proxy-server", WithTLSConfig(nodeConfig.NodeCertPath(), nodeConfig.NodeKeyPath())))
+
+	routes := []api.Route{
+		api.NewVersionHandler(),
+		controller.NewGetNodeRoute(db.Manager),
+		controller.NewAddUserRoute(nodeCtrl),
+		controller.NewInstallAppRoute(nodeCtrl),
+		controller.NewStartProcessHandler(nodeCtrl),
+		controller.NewMigrationRoute(nodeCtrl),
+	}
+
+	router := api.NewRouter(routes, log, nodeCtrl, nodeConfig)
+	apiServer := &http.Server{
+		Addr:        fmt.Sprintf(":%s", constants.DefaultPortHabitatAPI),
+		TLSConfig:   tlsConfig,
+		Handler:     router,
+		BaseContext: func(net.Listener) context.Context { return context.Background() },
+	}
+	url, err := url.Parse(fmt.Sprintf("http://localhost:%s", constants.DefaultPortHabitatAPI))
+	if err != nil {
+		log.Fatal().Err(fmt.Errorf("error parsing Habitat API URL: %v", err))
+	}
+	err = proxy.Rules.Add("Habitat API", &reverse_proxy.RedirectRule{
+		ForwardLocation: url,
+		Matcher:         "/habitat/api",
+	})
+	if err != nil {
+		log.Fatal().Err(fmt.Errorf("error adding Habitat API proxy rule: %v", err))
+	}
+	eg.Go(serveFn(apiServer, "api-server", WithTLSConfig(nodeConfig.NodeCertPath(), nodeConfig.NodeKeyPath())))
+
+	// Wait for either os.Interrupt which triggers ctx.Done()
+	// Or one of the servers to error, which triggers egCtx.Done()
+	select {
+	case <-egCtx.Done():
+		log.Err(egCtx.Err()).Msg("sub-service errored: shutting down Habitat")
+	case <-ctx.Done():
+		log.Info().Msg("Interrupt signal received; gracefully closing Habitat")
+		stop()
+	}
+
+	// Shutdown the API server
+	err = apiServer.Shutdown(context.Background())
+	if err != nil {
+		log.Err(err).Msg("error on api-server shutdown")
+	}
+	log.Info().Msg("Gracefully shutdown Habitat API server")
+
+	// Shutdown the proxy server
+	err = proxyServer.Shutdown(context.Background())
+	if err != nil {
+		log.Err(err).Msg("error on proxy-server shutdown")
+	}
+	log.Info().Msg("Gracefully shutdown Habitat proxy server")
+
+	// Wait for the go-routines to finish
+	err = eg.Wait()
+	if err != nil {
+		log.Err(err).Msg("received error on eg.Wait()")
+	}
+	log.Info().Msg("Finished!")
+}
+
+// config provided to http.Server.ListenAndServeTLS()
+type tlsConfig struct {
```

Comment:
Why can't we just use the same type used by the `tls.Config` type?

Reply:
These are the cert & key file names, which is different from what's in `tls.Config`/